docs/man/man1/library.1
.TH LIBRARY 1
.SH NAME
library \- send information requests to appropriate organization
.PD
.SH SYNOPSIS
.B library [-1234567] [request string]
.PD
.SH DESCRIPTION
.I library
sends document/information requests to the
appropriate organization within the Library Network.
It also handles a variety of requests for other AT&T organizations
(e.g., Engineering Information Services).
It requires the user to discriminate between seven classes of requests
as indicated by the menu it displays:
.sp
.nf
1)  Order by number - this includes nearly everything announced
	by the AT&T Library Network (e.g. TMs, bulletin items) or handled
	by the Engineering Information Services (e.g. J docs, CPSs).
2)  Order item not announced by the AT&T Library Network. Available:
	 Buy a copy of a book; Address labels; Technical Reports; 
	 Internal document by date (id unknown); Photocopies; 
	 Subscribe to a journal; foreign language services.
3)  Subscribe or unsubscribe to a bulletin (e.g. Mercury, CTP)
4)  Submit a database search. Examples of available databases:
	book catalog, internal documents, AT&T personnel, released papers
5)  Request human assistance / interaction. Type(s) available:
	 General AT&T Library Network assistance; Reference question.
6)  Request AT&T Library Network information/services. Examples:
	LINUS info, products/services descriptions, loan/reserve status
7)  Read AT&T Library Network email transmissions

.fi
Note that the first class includes anything announced by the Library Network operated by AT&T Bell Laboratories.  
.PP
The main menu level of 
.I library 
can be skipped by giving
the number of the desired option (1 - 7)
as the first parameter to the
\f2library\f1.
.PP
The secondary menu levels in
.I library
(presently in options 2, 4, 5 and 6)
can be bypassed by selecting the desired option as the second
parameter.
For example, to do order a photocopy, use the
command
.ce
library -2 -p
.PP
Similarly, if you want to do a search of the personnel database,
you can do the command line
.ce
library -4 -p waldstein, r k
.PP
Requests for objects that can be meaningfully described with one
line can be entered on the command line.
This includes the things orderable by options 1, 3, 4, and 6.
Note that library tries in this case to function with a minimum
of interaction.
For example, if you request a TM this way,
you will not be given a chance to enter remarks connected with the request.
.PP
When permitted, the command line requests can include more than one item
(presently only options 1 and 3 support this).
An example command line is
.ce
library -1 123456-851234-56tm  5d123 ad-123456 
.PP
In options 2, 5, and 6,
.I library
will prompt for a variety of information of varying complexity.
A period (.) at any point in this session will delete the 
request being entered.
Blank lines (just hit return) will cause optional information to
be left out of the request.
A line consisting of tilde e (\~e) will, when a long response is 
permitted, put you into an editor.
This editor defaults to ed(1).
However, if the environment variable EDITOR is set, the specified
editor is used.
A line consisting of tilde r (\~r) will, when a long response is 
permitted, read in the indicated file.
.PP
Option 7 is a misfit, in that it is primarily a reader, not a
request transmitter; although it does allow requesting items.
This option is intended for reading electronic transmissions
from the library network: primarily responses to option 4 search
requests and ASAP (specialized searches like electronic Mercury).
A convenient way to use option 7 is to invoke it via a pipe
from mail(1), mailx(1) or post(1).
.ce
 | 3 "library -7"
This assumes that mail message 3 consists of a search result.
The quote marks are required due to post(1) and mailx(1) syntax.
.PP
Break causes 
.I library
to exit
without sending any requests.
.PP
In general for more information about what the library command can
do, go into each option and enter a question mark.
This will cause a description of how the option works and
what it can request.
.PP
Several other pieces of information can be passed
to
.I library
to ease and improve its usage.
This information is looked up in a file called .lib (or the file indicated
by the LIBFILE shell variable).
It expects this file to contain lines of the form:
.sp
.ti 10
ID: \f2individual's PAN or Social Security Number\f1
.ti 10
libname: \f2individual's last name\f1
.ti 10
liblog: \f2name of log file\f1
.ti 10
libcntl: \f2control information\f1
.ti 10
reader: \f2reader control information\f1
.ti 10
liblocal: \f2control information\f1
.PP
If this file is not found or lines of this form are not found,
then 
.I library
prompts for name and ID (PAN or SS#).
.PP
This information can also be passed to 
.I library
as the shell parameters: LIBID, LIBNAME, LIBLOG, LIBCNTL, and LIBLOCAL.
.PP
.I library
keeps a log of requests sent via library if a line in the .lib
file exists giving a log file name, i.e., if you have in
your .lib file a line of the form
.ce
liblog: \f2name of log file\f1
.PP
.I library
keeps a log of requests in that file.
This file is created in a form that can be read and manipulated
by the 
.I mail
command.
To read or modify the log file, type
.ce
mail -f \f2name of log file\f1
.PP
.I library
creates the log file in your HOME directory unless the file name
given starts with a slash (/).
.I library
will automatically check option 1 requests
for duplicates in the log file.
.PP
The libcntl information is sent with the request to the program
that receives the requests for the library networks.
Control information containing the letter ``a''
will cause an acknowledgement to be mailed back to you that
your request has been received.
Control information of the form ``m\f2number\f1''will determine the
maximum number of items retrieved by a search request.
For example, a control line of the form
.sp
.ce
libcntl: am100
.PP
will cause requests to be acknowledged and a maximum of 100 retrieved
search items to be mailed back to you.
.PP
The reader control information is intended to let you personalize
the way
.I library
option 7 works for you.
Each letter after the colon indicates a different option turned
on or off.
Presently available are the following:
.RS
.TP 5
b
causes library -7 to leave a
blank line between records when more than one is displayed on the screen.
.TP 5
c
causes 
.I library
to confirm that you want the entered requests transmitted.
It does this at the end of the session, before finishing.
.TP 5
n
is an interesting features causing no introductory menu
of available announcements to be displayed.
The reader then goes straight into the first announcement to be read,
and moves directly from one announcement to the next, without
displaying the menu of those available at each stage.
.RE
.PP
The liblocal information is used to control the execution of 
.I library.
Presently the only meaningful control is \f2x\fP.
This causes library(1) to assume you are an expert and the
prompts are generally much shorter.
.PP
.I library
also uses your .lib file to save various repetitious responses
for its own use.
These will prevent you from having to duplicate responses.
.PD
.SH BUGS
.I library
checks upon input whether the request is reasonable.
New styles of request numbers require program modification
before they are valid.
.PD
.SH FILES
.PD 0
.TP 15
$HOME/.lib
This optional file contains a PAN and name for
.I library
to use.
.TP 15
/usr/lib/library/library.help
The help message displayed by library.
.PD
.SH "SEE ALSO"
mail(1), post(1), mailx(1)
.\"	@(#)library.1	1.12  11/26/90 RnD added


v10/cmd/qsnap/library.c
#include <stdio.h>
#include "defines.h"

static char Mode = 2;	/* remembers resolution for check in window() */

advance(how)
{
	if (how)
		qcmd("\020", 1);	/* enable frame advance */
	else
		qcmd("\017", 1);	/* disable frame advance */
	qpause();
}

shutter(how)
{
	if (how)
		qcmd("\033", 1);	/* open shutter */
	else
		qcmd("\032", 1);	/* close and advance film */
	qpause();
	qpause();
	qpause();
	qpause();
	qpause();
	qpause();
}

beep()
{
	qcmd("\021", 1);	/* beep */
	qpause();
}

loadluts(wbuf)		/* 8 bit wide lookup tables */
	char *wbuf;
{	char rbuf[768];
	register int i;

	qcmd("\000", 1);	/* load the rgb LUT's */
	qpause();
	qwrite(wbuf, 768);
	qpause();
	qcmd("\001", 1);	/* read back rgb LUT's */
	qpause();

	if (qread(rbuf, 768) != 768)
		return 0;
	else
		for(i = 0; i < 768; i++)
			if (rbuf[i] != wbuf[i])
				return 0;
	qpause();
	return 1;
}

load12luts(wbuf)	/* 12 bit wide lookup tables */
	short *wbuf;
{
	register int i;
	short rbuf[768];

	printf("loading 12 bit lookup tables\n"); fflush(stdout);
	qpause();
	qcmd("\094", 1);	/* load the rgb LUT's */
	qpause();
	qwrite((char *)wbuf, 1536);	/* sizeof(short)*768 */
	qpause();
	qcmd("\096", 1);	/* read back rgb LUT's */
	qpause();

	if (qread((char *)rbuf, 1536) != 1536)
		return 0;
	else
		for(i = 0; i < 768; i++)
			if (rbuf[i] != wbuf[i])
				return 0;
	qpause();
	return 1;
}

singlepass(how)
{	/* set single pass pixel mode */
	switch (how) {
	default     :
	case NEUTRAL:	qcmd("\002", 1); printf("singlepass neutral\n"); break;
	case RED    :	qcmd("\003", 1); printf("singlepass red\n"); break;
	case GREEN  :	qcmd("\004", 1); printf("singlepass green\n"); break;
	case BLUE   :	qcmd("\005", 1); printf("singlepass blue\n"); break;
	}
	qpause();
}

threepass()
{	/* set three pass pixel mode */
	qcmd("\006", 1);
	qpause();
	printf("3pass mode\n");
}

nocalibs()
{	/* disable automatic calibrations */
	qcmd("\037", 1);
	qpause(); qpause();
}

testpattern(which)
{
	if (which)
		qcmd("\012", 1);	/* test pattern 1 */
	else
		qcmd("\013", 1);	/* test pattern 0 */
	qpause();
}

linedelay(n)
	unsigned char n;
{	/* interline time delay - to match host speed to qcr scanrate */

	qcmd("\014", 1);
	qpause();
	qwrite(n, 1);		/* max delay n msec per line */
	qpause();
}

version(data)
	char *data;
{
	qcmd("\015", 1);
	qpause();
	qread(data, 14);
	qpause();
}

window(x, y, w, h)
{
	unsigned char buf[8];
	int maxsz = (Mode == 2)?1536:3072;	/* 1536 = 2*768 = 2*3/4*1024 */

	if (w+x > maxsz || h-y > maxsz)
	{	printf("error: bad image size\n");
		if (Mode == 2)
			printf("\tmax: -1024,  768, 1024,  -768\n");
		else
			printf("\tmax: -2048, 1536, 2048, -1536\n");
		printf("\thave: %4d, %4d, %4d, %4d\n", x, y, x+w, y-h);
		return;
	}
	buf[0] = (x>>8)&255; buf[1] = x&255;
	buf[2] = (w>>8)&255; buf[3] = w&255;
	buf[4] = (y>>8)&255; buf[5] = y&255;
	buf[6] = (h>>8)&255; buf[7] = h&255;
	qpause();
	qcmd("\016", 1);
	qpause();
	qwrite(buf, 8);
	qpause();
	printf("window (%dx%d): %d %d %d %d\n", w, h, x, y, x+w, y-h);
}

filmtype(n)
{
	if (n == NONE || n == TMAX_100)
	{	customluts(n);
		return;
	}
	qcmd("\043", 1);
	qpause();
	switch (n) {
	default:
	case LINEAR    : qwrite("\001", 1); printf("linear lut's\n"); break;
	case POLA559   : qwrite("\002", 1); printf("filmtype pola559\n"); break;
	case EKTA100   : qwrite("\003", 1); printf("filmtype ekat100\n"); break;
	case EKTA100_4k: qwrite("\004", 1); printf("filmtype ekta100_4k\n"); break;
	case POLA559_4k: qwrite("\005", 1); printf("filmtype pola559_4k\n"); break;
	case POLA809_4k: qwrite("\007", 1); printf("filmtype pola809_4k\n"); break;
	}
	qpause();
}

moduleno()
{	char c;

	qcmd("\045", 1);
	qpause();
	qread(&c, 1);
	qpause();
	switch (c) {
	case 240: printf("no module\n"); break;
	case  16: printf("35mm module\n"); break;
	case  32: printf("4x5\" module\n"); break;
	case  64: printf("8x10\" module\n"); break;
	default : printf("bad module number %d\n", c);
	}
	return c;	/*  (none, m35, m120, or m240) */
}

setreso(n)
{
	if (n == '2')
	{	qcmd("\077", 1);
		printf("2k mode set\n"); 
	} else if (n == '4')
	{	qcmd("\078", 1);
		printf("4k mode set \n"); 
	} else
	{	printf("resolution must be either 2 or 4\n");
		return;
	}
	qpause();
	fflush(stdout);
}

resolution()
{
	Mode = qreso();
	switch (Mode) {
	case 2:  printf("2k mode\n"); break;
	case 4:  printf("4k mode\n"); break;
	default: printf("qcr: bad resolution %d\n", Mode);
	}
	fflush(stdout);
	qpause();
	return Mode;
}

handshake(how)
{
	if (how)
		qcmd("\063", 1);	/* enable full handshake */
	else
		qcmd("\062", 1);	/* disable (default) */
	qpause();
}

rotate(how)
{
	if (how)
		qcmd("\067", 1);	/* enable 90 deg rotation */
	else
		qcmd("\066", 1);	/* disable (default) */
	qpause();
}

brightness(r,g,b,n)
{
	char buf[4]; int oval;

	buf[0] = r; buf[1] = g; buf[2] = b; buf[3] = n;
	qcmd("\055", 1); qpause();
	qwrite(buf, 4);  qpause();
}

getbright(nq)
	char *nq;
{	char buf[10]; int i;

	qcmd("\073", 1); qpause();
	qread(buf, 10);  qpause();
	for (i = 0; i < 9; i++)
		nq[i] = buf[i+1]&0xFF;
	qpause();
}

whaterror(n)
{	/* primary erro code, as shown on display lights:
		128	64	32	16
		  8	 4	 2	 1
	*/
	switch (n) {
	case   1:	printf("out of film\n"); break;
	case 129:	printf("unexpected command byte\n"); break;
	case 130:	printf("unexpected data byte\n"); break;
	case 131:	printf("illegal command byte\n"); break;
	case 132:	printf("illegal data byte\n"); break;
	case 133:	printf("command interrupt\n"); break;
	case 134:	printf("unimplemented function\n"); break;
	case 145:	printf("memory fault\n"); break;
	case 146:	printf("gpib interface fault\n"); break;
	case 147:	printf("crt beam calibration fault\n"); break;
	case 161:	printf("module fault\n"); break;
	case 162:	printf("filter wheel fault\n"); break;
	default:	printf("-bad primary code (%d)-\n", n); break;
	}
}

qerror()
{	char buf[8];

	qcmd("\044", 1);
	qpause();
	qread(buf, 8);
	whaterror(buf[0]&0xFF);
	qpause();
	printf("Error: pri=0x%02x sec=0x%02x", (buf[0]&0xFF), (buf[1]&0xFF));
	printf("\n\tH=0x%02x L=0x%02x ", (buf[2]&0xFF), (buf[3]&0xFF));
	printf("D=0x%02x \n\tE=0x%02x ", (buf[4]&0xFF), (buf[5]&0xFF));
	printf("B=0x%02x C=0x%02x\n", (buf[6]&0xFF), (buf[7]&0xFF));
}

qreso()
{	unsigned char buf[64];
	int i, j;

	qpause();
	qcmd("\074", 1);
	qpause();
	qpause();
	qread(buf, 1);
	qread(&buf[1], buf[0]);
	qpause();
	qpause();
	return (int) buf[18];
}

qstatus()
{	unsigned char buf[64];
	int i, j;

	qcmd("\074", 1);
	qpause();
	memset(buf, 0, 64);
	qread(buf, 1);
	printf("QCR status (%d bytes):\n", buf[0]);
	qread(&buf[1], buf[0]);
	for (i = 0; i < 38; i++)
	{	switch (i) {
		case  6: printf("\timage height: %d\n", qnum(buf[i],buf[i+1]));
			 i++; break;
		case  8: printf("\ty-offset    : %d\n", qnum(buf[i],buf[i+1]));
			 i++; break;
		case 10: printf("\timage  width: %d\n", qnum(buf[i],buf[i+1]));
			 i++; break;
		case 12: printf("\tx-offset    : %d\n", qnum(buf[i],buf[i+1]));
			 i++; break;
		case 26: printf("\tframe counter: %d\n", qnum(buf[i],buf[i+1]));
			 i++; break;
		case 36: printf("\tdrift value : %d\n", qnum(buf[i],buf[i+1]));
			 i++; break;
		default: whatstatus(i, buf[i]); break;
		}
	}
}

qnum(a, b)
	unsigned char a, b;
{	short n;

	n = b + a<<8;
	return (int)n;
}

whatstatus(n, m)
	unsigned char m;
{
	switch (n) {
	case  0: if (m) printf("\texternal luts loaded\n"); break;
	case  1: if (m) printf("\tneutral  luts loaded\n"); break;
	case  2: printf("\t%d msec interline time delay\n", m); break;
	case  3: if (m) printf("\tend of film\n"); break;
	case  4: printf("\tautomatic filter wheel %sabled\n",(m)?"dis":" en");break;
	case  5: printf("\tauto camera operation  %sabled\n",(m)?"dis":" en");break;
	case 14: printf("\tmodule no: %d\n", m); break;
	case 15: printf("\tautomatic calibration %sabled\n",(m)?"en":"dis");break;
	case 16: printf("\tautomatic warbling    %sabled\n",(m)?"dis":" en");break;
	case 17: printf("\trunlength black jumps %sabled\n",(m)?"dis":" en");break;
	case 18: printf("\tresolution: %dk\n", m); break;
	case 19: printf("\tmode was %sternally selected\n", (m)?"ex":"in");break;
	case 20: if (m) printf("\timage is 90 degrees rotated\n"); break;
	case 21: if (m) printf("\timage is mirrored\n"); break;
	case 22: if (m) printf("\tunbuffered raster mode\n"); break;
	case 23: if (m) printf("\treturn to no filter disabled\n"); break;
	case 24: if (m) printf("\tx and y coordinates are swapped\n"); break;
	case 25: if (m) printf("\tred repeat is enabled\n"); break;
	case 28: if (m) printf("\tfilm is in camera\n"); break;
	case 29: if (m) printf("\thalf frame mode enabled\n"); break;
	case 30: if (m) printf("\tbulk back is in use\n"); break;
	case 31: if (m) printf("\timage is 180 degrees rotated\n"); break;
	case 32: if (m) printf("\t12 bit lookup tables are loaded\n"); break;
	case 33: printf("\tinternal timing constant 1: %d\n", m); break;
	case 34: printf("\tinternal timing constant 2: %d\n", m); break;
	case 35: if (m) printf("\t'irradiance modulation' is active\n"); break;
	default: printf("\t%2d: %2d\n", n, m); break;
	}
}

/* customized lookup tables */

#include "qlut2.h"	/* 8b  generated by /usr/gerard/CAL/QCR/qlutnew.c */

/*
#include "LUT_pola52.h"	** awful, reversed? **
*/

unsigned char lutpola52[] = {	/* pola52 lut for brightness 148 */
	   0,   46,   54,   61,   67,   68,   68,   69,   69,   70, 
	  70,   71,   71,   72,   72,   73,   73,   74,   74,   75, 
	  75,   76,   76,   77,   77,   78,   78,   79,   79,   80, 
	  80,   81,   81,   82,   82,   83,   83,   84,   84,   85, 
	  85,   85,   86,   86,   86,   86,   87,   87,   87,   88, 
	  88,   88,   88,   89,   89,   89,   90,   90,   90,   90, 
	  91,   91,   91,   92,   92,   92,   92,   93,   93,   93, 
	  94,   94,   94,   94,   95,   95,   95,   96,   96,   96, 
	  96,   97,   97,   97,   98,   98,   98,   98,   99,   99, 
	  99,  100,  100,  100,  100,  101,  101,  101,  102,  102, 
	 103,  103,  104,  104,  105,  105,  105,  106,  106,  107, 
	 107,  108,  108,  109,  109,  110,  110,  111,  111,  112, 
	 112,  113,  113,  114,  114,  115,  115,  115,  116,  116, 
	 117,  117,  118,  118,  119,  119,  120,  120,  120,  121, 
	 121,  122,  122,  122,  123,  123,  124,  124,  124,  125, 
	 125,  126,  126,  126,  127,  127,  128,  128,  128,  129, 
	 129,  130,  130,  130,  131,  131,  132,  132,  132,  133, 
	 133,  134,  134,  134,  135,  135,  136,  136,  136,  137, 
	 137,  138,  138,  139,  139,  139,  140,  140,  141,  141, 
	 141,  142,  142,  143,  143,  144,  144,  144,  145,  145, 
	 146,  146,  147,  147,  147,  148,  148,  149,  149,  150, 
	 150,  150,  151,  151,  152,  152,  153,  154,  154,  155, 
	 156,  157,  158,  159,  160,  161,  161,  162,  163,  164, 
	 165,  166,  167,  168,  169,  169,  171,  173,  175,  176, 
	 178,  180,  182,  184,  185,  188,  195,  201,  205,  208, 
	 211,  215,  218,  221,  234,  255, 
};


customluts(n)
{	short all12uts[768];
	unsigned char alluts[768];
	register int i;

	switch (n) {
	case TMAX_100:
		for (i = 0; i < 768; i++)
			alluts[i] = qcr_lut[i%256];
		if (!loadluts(alluts))
		printf("loading new lookup tables failed\n");
		break;
	default:	/* polaroid type 52 */
		for (i = 0; i < 768; i++)
			alluts[i] = lutpola52[i%256];
		if (!loadluts(alluts))
		printf("loading new lookup tables failed\n");
		break;
	}
}

/*
not implemented:
18	 n	 +	red     runlength encoded single pass pixel mode
19	 n	 +	green   runlength sppm
20	 n	 +	blue    runlength sppm
21	 n	 +	neutral runlength sppm
22	 n	 +	three pass runlength image

24	 -	 -	disable automatic filter wheel control
25	 -	 -	enable  automatic filter wheel control (default)

29	256	 -	load a neutral LUT (default for neutral is blue lut)
30	256	 -	read neutral LUT
31/32	 -	 -	disable/enable automatic calibrations
33/34	 -	 -	disable/enable warble sounds at end of image
38	 -	 +	start an intensity calibration for used levels
39	 -	 +	intensity calibration for all 9 levels

40/41	 -	 -	disable/enable black jumping (for runlength)
48/49	 -	 -	disable/enable mirror imaging
52/53	 -	 -	disable/enable end-of-image filter wheel movement
54/55	 -	 -	disable/enable switch of x/y coordinates
56/57	 -	 -	disable/enable red repeat mode
*/
